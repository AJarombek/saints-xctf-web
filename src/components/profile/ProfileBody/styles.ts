/**
 * JSS styles for the ProfileBody component.
 * @author Andrew Jarombek
 * @since 9/7/2020
 */

import Colors from "../../../styles/colors";

export default {
    container: {
        minHeight: 'calc(100vh - 100px)',
        backgroundColor: Colors.lightBackground,
        margin: '100px 4% 0 4%',
        display: 'flex',

        '& > aside': {
            flexBasis: '30%',
            marginRight: 20
        },

        '& > section': {
            flexBasis: '70%'
        }
    },
    '@media screen and (max-width: 1200px)': {
        container: {
            margin: '100px 6% 0 2%',
        }
    },
    '@media screen and (max-width: 900px)': {
        container: {
            margin: '100px 2% 0 2%',
        }
    },
};
