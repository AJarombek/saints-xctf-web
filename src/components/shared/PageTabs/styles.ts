/**
 * JSS styles for the PageTabs component.
 * @author Andrew Jarombek
 * @since 9/8/2020
 */

import { FontMixins } from '../../../styles/mixins';
import Colors from '../../../styles/colors';
import color from 'color';

export default {
  tabs: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'flex-start',
    margin: '15px auto',
    width: 150,
  },
  tab: {
    textAlign: 'center',
    ...FontMixins.robotoSlab(),
    fontSize: 16,
    cursor: 'pointer',
    color: '#555',
    transition: 'color 0.5s ease',

    '&:hover': {
      color: '#222',
    },
  },
  currentTab: {
    color: Colors.sxctfRed,
    textDecoration: 'underline',
    textDecorationThickness: 2,
    textDecorationColor: '#888',
    transition: 'color 0.5s ease, text-decoration-color 0.5s ease',

    '&:hover': {
      color: color(Colors.sxctfRed).darken(0.1).hex(),
      textDecorationColor: '#444',
    },
  },
};
