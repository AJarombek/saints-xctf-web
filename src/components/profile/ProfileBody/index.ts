import ProfileBody from './ProfileBody';

export default ProfileBody;
