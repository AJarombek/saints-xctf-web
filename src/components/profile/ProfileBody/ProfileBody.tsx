/**
 * Component for the body of the user profile page.
 * @author Andrew Jarombek
 * @since 9/7/2020
 */

import React, { useEffect, useMemo, useState } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import PictureTitle from '../../shared/PictureTitle/PictureTitle';
import Flair from '../Flair/Flair';
import Memberships from '../Memberships/Memberships';
import PageTabs from '../../shared/PageTabs/PageTabs';
import {
  FlairMeta,
  LogFeeds,
  RangeViewExerciseTypeFilters,
  RootState,
  UserMeta,
  Users,
  StatsMeta
} from '../../../redux/types';
import PaginationBar from '../../shared/PaginationBar/PaginationBar';
import LogFeed from '../../shared/LogFeed/LogFeed';
import MonthlyCalendar from '../MonthlyCalendar';
import { logFeed } from '../../../redux/modules/logs';
import { useDispatch, useSelector } from 'react-redux';
import { getUserFlair, getUserMemberships } from '../../../redux/modules/profile';
import WeeklyChart from '../WeeklyChart';
import ProfileDetails from '../ProfileDetails';
import EditProfile from '../EditProfile';

interface Props {
  user: UserMeta;
  signedInUser: UserMeta;
  flair: FlairMeta;
  stats: StatsMeta;
  rangeViews: RangeViewExerciseTypeFilters;
}

export enum ProfileTab {
  LOGS,
  CALENDAR,
  CHART,
  DETAILS,
  EDIT
}

const useStyles = createUseStyles(styles);

const ProfileBody: React.FunctionComponent<Props> = ({ user, signedInUser, flair, stats, rangeViews }) => {
  const classes = useStyles();

  const dispatch = useDispatch();
  const userProfiles: Users = useSelector((state: RootState) => state.profile.users);
  const logFeeds: LogFeeds = useSelector((state: RootState) => state.logs.feeds);

  const [tab, setTab] = useState<ProfileTab>(ProfileTab.LOGS);
  const [bucket, setBucket] = useState<string>(null);
  const [page, setPage] = useState(1);

  useEffect(() => {
    if (bucket) {
      dispatch(logFeed('user', bucket, 10, 10 * (page - 1)));
    }
  }, [bucket, page, dispatch]);

  useEffect(() => {
    if (user) {
      if (user.username !== bucket) {
        setBucket(user.username);
      }

      if (!userProfiles[user.username]?.memberships && user.username) {
        dispatch(getUserMemberships(user.username));
      }

      if (!flair && user.username) {
        dispatch(getUserFlair(user.username));
      }
    }
  }, [user, userProfiles, flair, bucket, dispatch]);

  const totalPages: number = useMemo(() => {
    return logFeeds[`user-${bucket}`]?.pages[page]?.pages ?? 0;
  }, [logFeeds, page, bucket]);

  const tabs = useMemo(() => {
    const defaultTabs = [
      { tab: ProfileTab.LOGS, onView: (): void => setTab(ProfileTab.LOGS), content: 'Exercise Logs' },
      {
        tab: ProfileTab.CALENDAR,
        onView: (): void => setTab(ProfileTab.CALENDAR),
        content: 'Monthly Calendar'
      },
      { tab: ProfileTab.CHART, onView: (): void => setTab(ProfileTab.CHART), content: 'Weekly Chart' },
      { tab: ProfileTab.DETAILS, onView: (): void => setTab(ProfileTab.DETAILS), content: 'Details' }
    ];

    if (signedInUser && user?.username === signedInUser.username) {
      defaultTabs.push({ tab: ProfileTab.EDIT, onView: (): void => setTab(ProfileTab.EDIT), content: 'Edit Profile' });
    }

    return defaultTabs;
  }, [user, signedInUser]);

  if (user) {
    return (
      <div className={classes.container}>
        <aside>
          <PictureTitle
            imageUrl={
              user?.profilepic_name
                ? `/uasset/profile/${user?.username}/${user?.profilepic_name}`
                : '/asset/saintsxctf.png'
            }
            title={`${user?.first} ${user?.last}`}
            subTitle={`@${user?.username}`}
          />
          <Flair flair={flair} />
          <Memberships teamMemberships={userProfiles[user.username]?.memberships} />
          <PageTabs currentTab={tab} tabs={tabs} />
        </aside>
        <section>
          {tab === ProfileTab.LOGS && (
            <>
              <LogFeed logFeeds={logFeeds} page={page} user={signedInUser} filterBy="user" bucket={bucket} />
              <PaginationBar
                page={page}
                totalPages={totalPages}
                onChangePage={(page): void => {
                  setPage(page);
                  window.scrollTo(0, 0);
                }}
              />
            </>
          )}
          {tab === ProfileTab.CALENDAR && <MonthlyCalendar rangeViews={rangeViews} user={user} />}
          {tab === ProfileTab.CHART && <WeeklyChart rangeViews={rangeViews} user={user} />}
          {tab === ProfileTab.DETAILS && <ProfileDetails user={user} stats={stats} />}
          {tab === ProfileTab.EDIT && user?.username === signedInUser?.username && <EditProfile user={user} />}
        </section>
      </div>
    );
  } else {
    return null;
  }
};

export default ProfileBody;
