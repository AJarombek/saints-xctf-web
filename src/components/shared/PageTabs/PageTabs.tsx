/**
 * Component for on the profile and group pages which lists all the viewable tabs.
 * @author Andrew Jarombek
 * @since 9/8/2020
 */

import React, { ReactNode } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import { ProfileTab } from '../../profile/ProfileBody/ProfileBody';
import classNames from 'classnames';

type Tab = {
  tab: number;
  onView: () => void;
  content: ReactNode;
};

interface Props {
  currentTab: number;
  tabs: Tab[];
}

const useStyles = createUseStyles(styles);

const PageTabs: React.FunctionComponent<Props> = ({ currentTab, tabs }) => {
  const classes = useStyles();

  return (
    <div className={classNames(classes.tabs, 'tabs')}>
      {tabs.map((tab: Tab) => (
        <p
          className={classNames(classes.tab, currentTab === tab.tab && classes.currentTab)}
          onClick={tab.onView}
          key={tab.tab}
        >
          {tab.content}
        </p>
      ))}
    </div>
  );
};

export default PageTabs;
