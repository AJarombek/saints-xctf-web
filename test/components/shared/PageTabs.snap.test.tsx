/**
 * Snapshot test for the PageTabs component.
 * @author Andrew Jarombek
 * @since 3/7/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import PageTabs from '../../../src/components/shared/PageTabs';
import { ProfileTab } from '../../../src/components/profile/ProfileBody/ProfileBody';

describe('PageTabs Snapshot Tests', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <PageTabs
          currentTab={ProfileTab.LOGS}
          tabs={[
            { tab: ProfileTab.LOGS, onView: (): void => null, content: 'Exercise Logs' },
            { tab: ProfileTab.CALENDAR, onView: (): void => null, content: 'Monthly Calendar' },
            { tab: ProfileTab.CHART, onView: (): void => null, content: 'Weekly Chart' },
            { tab: ProfileTab.DETAILS, onView: (): void => null, content: 'Details' }
          ]}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
