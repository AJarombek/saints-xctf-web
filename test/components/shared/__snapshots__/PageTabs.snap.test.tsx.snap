// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTabs Snapshot Tests renders correctly 1`] = `
<div
  className="tabs-0-1-1 tabs"
>
  <p
    className="tab-0-1-2 currentTab-0-1-3"
    onClick={[Function]}
  >
    Exercise Logs
  </p>
  <p
    className="tab-0-1-2"
    onClick={[Function]}
  >
    Monthly Calendar
  </p>
  <p
    className="tab-0-1-2"
    onClick={[Function]}
  >
    Weekly Chart
  </p>
  <p
    className="tab-0-1-2"
    onClick={[Function]}
  >
    Details
  </p>
</div>
`;
